torch/_dynamo/config.py

```diff
@@ -381,6 +381,28 @@ def _get_optimize_ddp_mode():
 # torch._dynamo.utilsCompileTimeInstructionCounter.
 record_compile_time_instruction_count = False
 
+# Number of eager warmup runs to run before compiling.
+# Example:
+# ```
+# torch._dynamo.config.warmup_runs = 1
+# torch._dynamo.config.compiled_autograd = True
+#
+# @torch.compile
+# def fn(inputs):
+#     output = model(inputs)
+#     loss = loss_fn(output)
+#     loss.backward()
+#
+# for i in range(3):
+#     # 1st iteration: forward is run in eager, backward is run in eager.
+#     # 2nd iteration and beyond: forward is run with compile, backward is run with Compiled Autograd.
+#     fn(inputs)
+# ```
+# NOTE: if you are using the standalone Compiled Autograd context manager,
+# to have warmup runs for Compiled Autograd you must reuse the same context manager for all iterations.
+# (This is because we do the stateful tracking of the warmup count within that context manager.)
```

Comment:
This admonition makes me a bit suspicious about the API model you have for warmup. Right now, you are counting warmup iterations inside of context managers / decorators like torch.compile. But this can lead to some pretty strange behavior. The one here is one case, you can potentially also suffer from the same problem with torch.compile if that decorator is not durable, and you can also have some funny business when you have multiple torch.compile regions but you don't hit them all at the same rate, and so some of them get out of warmup before the others.

Like, what if you just had a different API, a single function call, which puts you out of warmup mode globally? Then there would be none of this faffing about; you warmup, you call a function, and now you are not warmup anymore.

Reply:
Thanks and agreed it's an important concern. For this API usage pattern: "you warmup, you call a function, and now you are not warmup anymore", wonder would it mean that the user training loop would be something like:
```python
for i in range(num_iters):
    out = maybe_warmup(torch.compile(fn), num_runs=1)(x)   # run everything in eager (warmup mode) 
    if iter > config.warmup_runs:
        stop_warmup()  # after this, turn off warmup and run everything in compile
```

Reply:
Yes, exactly, my point is that in complex cases, it seems to me that it is better for the training loop to do it, much simpler and intuitive for users.

Reply:
One idea is:
```python
for i in range(num_iters):
    out = torch.compile(fn, options={"warmup": 1})(x)
```
and under the hood do the warmup count tracking on an attribute of `fn` (e.g. `fn._torchdynamo_warmup_count`)

If user's `fn` is ephemeral (created per training loop), I feel that it's probably an edge case and we can warn user about this whenever `warmup` option is set.

Reply:
But you are still out of luck for the compiled autograd context manager case, no?

Reply:
The compiled autograd context manager case will also be an edge use case - we want to advocate for using `torch._dynamo.config.compiled_autograd=True` instead, which will have a persistent compiled autograd context manager underneath, and user will not need a standalone CA context manager.

I feel that we still need this warmup API because otherwise it limits at which step in the training setup code can the user apply compile. i.e. if user cannot apply compile until 2nd iteration, it means `torch.compile()` must be applied last among a list of things (e.g. activation checkpoint, tensor parallel, FSDP2, pipeline parallel, torch.compile), which I feel would be too limiting and might cause user's training loop code to not be very clean.

Reply:
> Is it that if torch.compile closes over a backward() call, compiled autograd will automatically be enabled? Is there a PR with this proposal somewhere?

Yes provided the config flag is turned on on, https://github.com/pytorch/pytorch/pull/125880

Reply:
> I mean, what if it is literally the stop_warmup API call you proposed? Here, torch.compile can be anywhere in the stack, and it's just this function call that modulates between warmup versus not.

sg - I can go in this direction, with this proposed API:
```python
torch._dynamo.config.compiled_autograd = True

@torch.compile
def iter_fn(x):
    out = fn(x)
    loss = loss_fn(out)
    loss.backward()
    return out, loss

warmup_runs = 1
torch._dynamo.utils.start_warmup()  # after this, turn on warmup and run everything in eager
for i in range(num_iters):
    out, loss = iter_fn(x)
    if i > warmup_runs:
        torch._dynamo.utils.stop_warmup()  # after this, turn off warmup and run everything in compile
```